src/media/js/addon/actions/submitVersion.js

```diff
@@ -0,0 +1,131 @@
+import req from 'request';
+import Url from 'urlgray';
+import urlJoin from 'url-join';
+
+
+export const VALIDATION_BEGIN = 'ADDON_SUBMIT_VERSION__VALIDATION_BEGIN';
+const validationBegin = createAction(VALIDATION_BEGIN);
+
+// For when the add-on upload has started processing.
+export const VALIDATION_PENDING = 'ADDON_SUBMIT_VERSION__VALIDATION_PENDING';
+const validationPending = createAction(VALIDATION_PENDING);
+
+// For when the add-on upload has finished validation.
+export const VALIDATION_PASS = 'ADDON_SUBMIT_VERSION__VALIDATION_PASS';
+const validationPassed = createAction(VALIDATION_PASS);
+
+export const VALIDATION_FAIL = 'ADDON_SUBMIT_VERSION__VALIDATION_FAIL';
+const validationFail = createAction(VALIDATION_FAIL);
+
+export const SUBMIT_OK = 'ADDON_SUBMIT_VERSION__SUBMIT_OK';
+const submitOk = createAction(SUBMIT_OK);
+
+export const SUBMIT_ERROR = 'ADDON_SUBMIT_VERSION__SUBMIT_ERROR';
+const submitError = createAction(SUBMIT_ERROR);
+
+
+export function submitVersion(fileData, addonSlug) {
+  /*
+    Upload add-on .zip file for validation.
+
+    addonSlug -- will be passed down all the way from validation, polling, to
+                 actual version creation.
+  */
+  return (dispatch, getState) => {
+    const apiArgs = getState().apiArgs || {};
+    const validationUrl = Url(
+      urlJoin(process.env.MKT_API_ROOT, 'extensions/validation/')
+    ).q(apiArgs);
+
+    dispatch(validationBegin());
+
+    req
+      .post(validationUrl)
+      .set('Content-Type', 'application/zip')
+      .set('Content-Disposition',
+           'form-data; name="binary_data"; filename="extension.zip"')
+      .send(fileData)
+      .then(res => {
+        // Notify the reducers.
+        dispatch(validationPending(res.body.id));
+
+        // Take further action by polling.
+        dispatch(pollValidator(res.body.id, addonSlug));
+      }, err => {
+        dispatch(validationFail(JSON.parse(err.response.text).detail));
+      });
+  };
+}
+
+
+export function pollValidator(validationId, addonSlug) {
+  /*
+    Poll add-on's validation detail endpoint to wait until validation has
+    processed.
+  */
+  return (dispatch, getState) => {
+    const apiArgs = getState().apiArgs || {};
+    const pollUrl = Url(
+      urlJoin(process.env.MKT_API_ROOT,
+              `extensions/validation/${validationId}/`)
+    ).q(apiArgs);
+
+    function poll(interval) {
+      req
+        .get(pollUrl)
+        .then(res => {
+          if (res.body.processed) {
+            // If processed, then validation is complete. Resolve process.
+            clearInterval(pollValidatorInterval);
+
+            if (res.body.valid) {
+              dispatch(createVersion(addonSlug));
+            } else {
+              dispatch(validationFail(res.body.validation));
+            }
+          }
+        });
+    }
+
+    // Poll the validator, poll() will clear interval when complete + dispatch.
+    const pollValidatorInterval = setInterval(() => {
```

Comment:
Does the validation clearer actually have access to the object? I thought with ES6 scoping it wouldn't. I'd double-check to make sure you're not no-opping.